File: software/experimental/wxRuby/notifications.rb
class Notification

    @@observers = {}

    def initialize(value)
        @value = value
    end

    def Notification.add_observer(callback)
        if not @@observers.has_key?(self.__id__)
            @@observers[self.__id__] = []
        end
        @@observers[self.__id__] << callback
    end

    def Notification.remove_observer(callback)
        @@observers.delete(callback)
    end

    def Notification.notify(value)
        if @@observers.has_key?(self.__id__)
            @@observers[self.__id__].each do |callback|
                callback.call(new(value))
            end
        end
    end
end

class KeyUpdateNotification < Notification
    def eid; @value; end
end

class LayerChangeNotification < Notification
    def layer; @value; end
end

class ModelChangeNotification < Notification
    def keyboardModel; @value; end
end

class CategoryChangeNotification < Notification
    def category; @value; end
end

if $0 == __FILE__

    class SomeOther < Notification
        def val; @value; end
    end

    class Test

        def initialize()
            KeyUpdateNotification.add_observer(self.method(:OnKeyUpdate))
            SomeOther.add_observer(self.method(:OnSomeOther))
        end

        def OnKeyUpdate(note)
            puts note.eid
        end

        def OnSomeOther(note)
            puts note.val
        end
    end

    test = Test.new()

    KeyUpdateNotification.notify(100)
    puts
    SomeOther.notify(200)

end

File: software/experimental/wxPython/notifications.py
class Notification(object):

    @classmethod
    def AddObserver(cls, observerFunc):
        cls._observers.append(observerFunc)

    @classmethod
    def RemoveObserver(cls, observerFunc):
        cls._observers.remove(observerFunc)

    def notify(self):
        for updateFunc in self.__class__._observers:
            updateFunc(self)

class KeyUpdateNotification(Notification):

    _observers = []
    def __init__(self, eid):
        self.eid = eid

class LayerChangeNotification(Notification):

    _observers = []
    def __init__(self, layer):
        self.layer = layer

class ModelChangeNotification(Notification):

    _observers = []
    def __init__(self, keyboardModel):
        self.keyboardModel = keyboardModel

class CategoryChangeNotification(Notification):

    _observers = []
    def __init__(self, category):
        self.category = category

def main():

    class Test:
        def __init__(self):
            KeyUpdateNotification.AddObserver(self.OnKeyUpdate)
            SomeOther.AddObserver(self.OnSomeOther)

        def OnKeyUpdate(self, note):
            print note.eid

        def OnSomeOther(self, note):
            print note.val

    class SomeOther(Notification):

        _observers = []
        def __init__(self, val):
            Notification.__init__(self)
            self.val = val

    test = Test()

    notification = KeyUpdateNotification(100)
    notification.notify()
    print
    notification = SomeOther(200)
    notification.notify()

if __name__ == '__main__':
    main()




